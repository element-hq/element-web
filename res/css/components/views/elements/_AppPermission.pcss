/*
Copyright 2024 New Vector Ltd.
Copyright 2019-2023 The Matrix.org Foundation C.I.C.
Copyright 2015, 2016 OpenMarket Ltd

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_AppPermission {
    font-size: $font-12px;
    width: 100%; /* make mx_AppPermission fill width of mx_AppTileBody so that scroll bar appears on the edge */
    overflow-y: scroll;
    .mx_AppPermission_bolder {
        font-weight: var(--cpd-font-weight-semibold);
    }
    .mx_AppPermission_content {
        margin-block: auto; /* place at the center */

        > div {
            margin-block: 12px;
        }

        .mx_AppPermission_content_bolder {
            font-weight: var(--font-semi-bold);
        }

        .mx_TextWithTooltip_target--helpIcon {
            display: inline-block;
            height: $font-14px; /* align with characters on the same line */
            vertical-align: middle;

            .mx_Icon {
                color: $accent;
            }
        }
    }
}
