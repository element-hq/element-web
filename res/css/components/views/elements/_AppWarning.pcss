/*
Copyright 2024 New Vector Ltd.
Copyright 2023 Suguru Hirahara

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_AppWarning {
    font-size: $font-16px;
    justify-content: center;

    h4 {
        margin: 0;
        padding: 0;
    }
}
