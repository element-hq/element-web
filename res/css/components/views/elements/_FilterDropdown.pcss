/*
Copyright 2024 New Vector Ltd.
Copyright 2022 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_FilterDropdown {
    .mx_Dropdown_menu {
        margin-top: $spacing-4;
        left: unset;
        right: -$spacing-12;
        width: 232px;
        padding: $spacing-12;

        border: 1px solid $quinary-content;
        border-radius: 8px;
        box-shadow: 0px 1px 3px rgba(23, 25, 28, 0.05);

        background-color: $system;

        .mx_Dropdown_option_highlight {
            background-color: transparent;
        }
    }

    .mx_Dropdown_input {
        height: 24px;
        background-color: transparent;
        border-color: transparent;
        color: $secondary-content;
        border-radius: 4px;

        &:focus,
        &:hover {
            background-color: $quinary-content;
            border-color: $quinary-content;
        }
    }

    .mx_Dropdown_arrow {
        background: $secondary-content;
    }
}

.mx_FilterDropdown_option {
    position: relative;
    width: 100%;
    box-sizing: border-box;
    padding: $spacing-8 0 $spacing-8 $spacing-20;

    font-size: $font-12px;
    line-height: $font-15px;
    color: $primary-content;
}

.mx_FilterDropdown_optionSelectedIcon {
    height: 14px;
    width: 14px;
    position: absolute;
    top: $spacing-8;
    left: 0;
}

.mx_FilterDropdown_optionLabel {
    font-weight: var(--cpd-font-weight-semibold);
    display: block;
}

.mx_FilterDropdown_optionDescription {
    color: $secondary-content;
    margin-top: $spacing-4;
}
