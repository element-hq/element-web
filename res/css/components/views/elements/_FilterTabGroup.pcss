/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_FilterTabGroup {
    color: $primary-content;
    label {
        margin-right: $spacing-12;
        cursor: pointer;
        span {
            display: inline-block;
            line-height: $font-24px;
        }
    }
    input[type="radio"] {
        appearance: none;
        margin: 0;
        padding: 0;

        &:focus,
        &:hover {
            & + span {
                color: $secondary-content;
            }
        }

        &:checked + span {
            color: $accent;
            font-weight: var(--cpd-font-weight-semibold);
            /* underline */
            box-shadow: 0 1.5px 0 0 currentColor;
        }
    }
}
