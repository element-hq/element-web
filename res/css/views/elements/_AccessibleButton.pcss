/*
Copyright 2017 Vector Creations Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_AccessibleButton {
    cursor: pointer;

    &.mx_AccessibleButton_disabled {
        cursor: not-allowed;

        &.mx_AccessibleButton_kind_icon_primary,
        &.mx_AccessibleButton_kind_icon_primary_outline,
        &.mx_AccessibleButton_kind_primary,
        &.mx_AccessibleButton_kind_primary_outline,
        &.mx_AccessibleButton_kind_primary_sm,
        &.mx_AccessibleButton_kind_link,
        &.mx_AccessibleButton_kind_link_accent,
        &.mx_AccessibleButton_kind_link_inline,
        &.mx_AccessibleButton_kind_danger_inline,
        &.mx_AccessibleButton_kind_content_inline,
        &.mx_AccessibleButton_kind_link_sm {
            opacity: 0.4;
        }
    }

    &.mx_AccessibleButton_hasKind {
        padding: 7px 18px;
        text-align: center;
        border-radius: 24px;
        display: inline-flex;
        align-items: center;
        justify-content: center;
        font: var(--cpd-font-body-md-semibold);
        border: none; /* override default <button /> styles */
        word-break: keep-all; /* prevent button text in Chinese/Japanese/Korean (CJK) from being collapsed */

        &.mx_AccessibleButton_kind_primary_sm,
        &.mx_AccessibleButton_kind_danger_sm,
        &.mx_AccessibleButton_kind_link_sm {
            padding: 5px 12px;
        }

        &.mx_AccessibleButton_kind_primary_sm {
            color: var(--cpd-color-text-on-solid-primary);
            background-color: var(--cpd-color-bg-action-primary-rest);
        }

        &.mx_AccessibleButton_kind_danger_sm {
            color: var(--cpd-color-text-on-solid-primary);
            background-color: var(--cpd-color-bg-critical-primary);
        }

        &.mx_AccessibleButton_kind_link_sm {
            color: var(--cpd-color-text-primary);
            text-decoration: underline;
            font-weight: var(--cpd-font-weight-semibold);
        }

        &.mx_AccessibleButton_kind_confirm_sm {
            background-color: var(--cpd-color-bg-action-primary-rest);

            &::before {
                mask-image: url("$(res)/img/feather-customised/check.svg");
            }
        }

        &.mx_AccessibleButton_kind_cancel_sm {
            background-color: var(--cpd-color-bg-critical-primary);

            &::before {
                mask-image: url("$(res)/img/feather-customised/x.svg");
            }
        }

        &.mx_AccessibleButton_kind_icon,
        &.mx_AccessibleButton_kind_icon_primary,
        &.mx_AccessibleButton_kind_icon_primary_outline {
            padding: 0;
            height: 32px;
            width: 32px;
        }
    }

    &.mx_AccessibleButton_kind_icon_primary,
    &.mx_AccessibleButton_kind_icon_primary_outline,
    &.mx_AccessibleButton_kind_primary,
    &.mx_AccessibleButton_kind_primary_outline,
    &.mx_AccessibleButton_kind_secondary {
        font-weight: var(--cpd-font-weight-semibold);
    }

    &.mx_AccessibleButton_kind_icon_primary,
    &.mx_AccessibleButton_kind_primary {
        border: 1px solid var(--cpd-color-bg-action-primary-rest);
        color: var(--cpd-color-text-on-solid-primary);
        background-color: var(--cpd-color-bg-action-primary-rest);
    }

    &.mx_AccessibleButton_kind_icon_primary_outline,
    &.mx_AccessibleButton_kind_primary_outline {
        border: 1px solid var(--cpd-color-border-interactive-secondary);
        color: var(--cpd-color-text-primary);
    }

    &.mx_AccessibleButton_kind_secondary {
        color: var(--cpd-color-text-primary);
        text-decoration: underline;
    }

    &.mx_AccessibleButton_kind_secondary_content {
        color: $secondary-content;
    }

    &.mx_AccessibleButton_kind_danger {
        color: var(--cpd-color-text-on-solid-primary);
        background-color: var(--cpd-color-bg-critical-primary);

        &.mx_AccessibleButton_disabled {
            color: var(--cpd-color-text-on-solid-primary);
            background-color: var(--cpd-color-bg-critical-primary);
        }
    }

    &.mx_AccessibleButton_kind_danger_outline {
        color: var(--cpd-color-text-critical-primary);
        background-color: transparent;
        border: 1px solid var(--cpd-color-border-critical-subtle);

        &.mx_AccessibleButton_disabled {
            color: var(--cpd-color-text-disabled);
            border-color: var(--cpd-color-border-disabled);
        }
    }

    &.mx_AccessibleButton_kind_danger_sm {
        &.mx_AccessibleButton_disabled {
            color: var(--cpd-color-text-disabled);
            background-color: var(--cpd-color-bg-subtle-primary);
        }
    }

    &.mx_AccessibleButton_kind_link,
    &.mx_AccessibleButton_kind_link_inline,
    &.mx_AccessibleButton_kind_danger_inline,
    &.mx_AccessibleButton_kind_content_inline {
        font-size: inherit;
        font-weight: var(--cpd-font-weight-semibold);
        line-height: inherit;
        padding: 0;
        text-decoration: underline;
    }

    &.mx_AccessibleButton_kind_link,
    &.mx_AccessibleButton_kind_link_inline {
        color: var(--cpd-color-text-primary);
    }

    &.mx_AccessibleButton_kind_danger_inline {
        color: var(--cpd-color-text-critical-primary);
    }

    &.mx_AccessibleButton_kind_content_inline {
        color: $primary-content;
    }

    &.mx_AccessibleButton_kind_link_inline,
    &.mx_AccessibleButton_kind_danger_inline,
    &.mx_AccessibleButton_kind_content_inline {
        display: inline;
    }

    &.mx_AccessibleButton_kind_confirm_sm,
    &.mx_AccessibleButton_kind_cancel_sm {
        padding: 0px;
        width: 16px;
        height: 16px;
        border-radius: 100%;
        position: relative;
        display: block;

        &::before {
            content: "";
            display: block;
            position: absolute;
            inset: 0;
            background-color: #ffffff;
            mask-repeat: no-repeat;
            mask-position: center;
            mask-size: 80%;
        }
    }
}
