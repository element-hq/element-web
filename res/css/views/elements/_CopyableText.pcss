/*
Copyright 2024 New Vector Ltd.
Copyright 2022 Å imon Brandner <simon.bra.ag@gmail.com>
Copyright 2019 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_CopyableText {
    align-items: flex-start;
    display: flex;
    justify-content: space-between;
    width: max-content;
    max-width: 100%;

    &.mx_CopyableText_border {
        overflow: auto;
        border-radius: 5px;
        border: solid 1px $light-fg-color;
        margin-bottom: 10px;
        margin-top: 10px;
        padding: 10px 0 10px 10px;
    }

    .mx_CopyableText_copyButton {
        flex-shrink: 0;
        /* using em here to adapt to the local font size */
        width: 1em;
        height: 1em;
        cursor: pointer;
        padding-left: 12px;
        padding-right: 10px;
        display: block;
        /* If the copy button is used within a scrollable div, make it stick to the right while scrolling */
        position: sticky;
        right: 0;
        /* center to first line */
        top: 0.15em;
        background-color: $background;

        &::before {
            content: "";
            mask-image: url($copy-button-url);
            mask-position: center center;
            mask-repeat: no-repeat;
            mask-size: contain;
            background-color: $message-action-bar-fg-color;
            width: 1em;
            height: 1em;
            display: block;
            background-repeat: no-repeat;
        }
    }
}
