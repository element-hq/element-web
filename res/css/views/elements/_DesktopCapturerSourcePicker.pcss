/*
Copyright 2024 New Vector Ltd.
Copyright 2021 Å imon Brandner <simon.bra.ag@gmail.com>

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

.mx_desktopCapturerSourcePicker {
    overflow: hidden;

    .mx_desktopCapturerSourcePicker_tab {
        display: flex;
        flex-wrap: wrap;
        justify-content: center;
        align-items: flex-start;
        height: 500px;
        overflow: overlay;

        .mx_desktopCapturerSourcePicker_source {
            width: 50%;
            display: flex;
            flex-direction: column;

            .mx_desktopCapturerSourcePicker_source_thumbnail {
                margin: 4px;
                padding: 4px;
                border-width: 2px;
                border-radius: 8px;
                border-style: solid;
                border-color: transparent;

                &.mx_desktopCapturerSourcePicker_source_thumbnail_selected,
                &:hover,
                &:focus {
                    border-color: $accent;
                }
            }

            .mx_desktopCapturerSourcePicker_source_name {
                margin: 0 4px;
                white-space: nowrap;
                text-overflow: ellipsis;
                overflow: hidden;
            }
        }
    }
}
