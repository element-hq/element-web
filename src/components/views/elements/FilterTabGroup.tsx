/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React, { FieldsetHTMLAttributes, ReactNode } from "react";

export type FilterTab<T> = {
    label: string | ReactNode;
    id: T;
};
type FilterTabGroupProps<T extends string = string> = FieldsetHTMLAttributes<any> & {
    // group name used for radio buttons
    name: string;
    onFilterChange: (id: T) => void;
    // active tab's id
    value: T;
    // tabs to display
    tabs: FilterTab<T>[];
};

/**
 * React component which styles a set of content filters as tabs
 *
 * This is used in displays which show a list of content items, and the user can select between one of several
 * filters for those items. For example, in the Poll History dialog, the user can select between "Active" and "Ended"
 * polls.
 *
 * Type `T` is used for the `value` attribute for the buttons in the radio group.
 */
export const FilterTabGroup = <T extends string = string>({
    name,
    value,
    tabs,
    onFilterChange,
    ...rest
}: FilterTabGroupProps<T>): JSX.Element => (
    <fieldset {...rest} className="mx_FilterTabGroup">
        {tabs.map(({ label, id }) => (
            <label data-testid={`filter-tab-${name}-${id}`} key={id}>
                <input type="radio" name={name} value={id} onChange={() => onFilterChange(id)} checked={value === id} />
                <span>{label}</span>
            </label>
        ))}
    </fieldset>
);
