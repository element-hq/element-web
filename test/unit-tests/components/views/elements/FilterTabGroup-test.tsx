/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { fireEvent, render } from "jest-matrix-react";

import { FilterTabGroup } from "../../../../../src/components/views/elements/FilterTabGroup";

describe("<FilterTabGroup />", () => {
    enum TestOption {
        Apple = "Apple",
        Banana = "Banana",
        Orange = "Orange",
    }
    const defaultProps = {
        "name": "test",
        "value": TestOption.Apple,
        "onFilterChange": jest.fn(),
        "tabs": [
            { id: TestOption.Apple, label: `Label for ${TestOption.Apple}` },
            { id: TestOption.Banana, label: `Label for ${TestOption.Banana}` },
            { id: TestOption.Orange, label: `Label for ${TestOption.Orange}` },
        ],
        "data-testid": "test",
    };
    const getComponent = (props = {}) => <FilterTabGroup<TestOption> {...defaultProps} {...props} />;

    it("renders options", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("calls onChange handler on selection", () => {
        const onFilterChange = jest.fn();
        const { getByText } = render(getComponent({ onFilterChange }));

        fireEvent.click(getByText("Label for Banana"));

        expect(onFilterChange).toHaveBeenCalledWith(TestOption.Banana);
    });
});
