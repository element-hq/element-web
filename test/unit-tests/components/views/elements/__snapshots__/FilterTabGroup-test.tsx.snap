// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterTabGroup /> renders options 1`] = `
<div>
  <fieldset
    class="mx_FilterTabGroup"
    data-testid="test"
  >
    <label
      data-testid="filter-tab-test-Apple"
    >
      <input
        checked=""
        name="test"
        type="radio"
        value="Apple"
      />
      <span>
        Label for Apple
      </span>
    </label>
    <label
      data-testid="filter-tab-test-Banana"
    >
      <input
        name="test"
        type="radio"
        value="Banana"
      />
      <span>
        Label for Banana
      </span>
    </label>
    <label
      data-testid="filter-tab-test-Orange"
    >
      <input
        name="test"
        type="radio"
        value="Orange"
      />
      <span>
        Label for Orange
      </span>
    </label>
  </fieldset>
</div>
`;
